File: storage/models/test_protocol.go
package models

import (
	"time"
)

type TestProtocol struct {
	Model
	TestRunID            uint                       `json:"testRunId,omitempty"`
	ParentTestProtocolID *uint                      `sql:"default:NULL" json:"parentTestProtocolId,omitempty"`
	TestRun              *TestRun                   `json:"testRun,omitempty"`
	DeviceID             *uint                      `json:"deviceId,omitempty"`
	Device               *Device                    `json:"device,omitempty"`
	TestName             string                     `json:"testName"`
	StartedAt            time.Time                  `json:"startedAt,omitempty"`
	EndedAt              *time.Time                 `json:"endedAt,omitempty"`
	Entries              []ProtocolEntry            `json:"entries,omitempty"`
	TestResult           TestResultState            `json:"testResult,omitempty"`
	Performance          []ProtocolPerformanceEntry `json:"performance,omitempty"`
	AvgFPS               float64                    `sql:"type:decimal(10,2);" json:"avgFps,omitempty"`
	AvgMEM               float64                    `sql:"type:decimal(10,2);" json:"avgMem,omitempty"`
	AvgCPU               float64                    `sql:"type:decimal(10,2);" json:"avgCpu,omitempty"`
	AvgVertexCount       float64                    `sql:"type:decimal(10,2);" json:"avgVertexCount,omitempty"`
	AvgTriangles         float64                    `sql:"type:decimal(10,2);" json:"avgTriangles,omitempty"`
	// calculated fields
	HistAvgFPS          float64        `db:"-" gorm:"-:all" json:"histAvgFps,omitempty"`
	HistAvgMEM          float64        `db:"-" gorm:"-:all" json:"histAvgMem,omitempty"`
	HistAvgCPU          float64        `db:"-" gorm:"-:all" json:"histAvgCpu,omitempty"`
	HistAvgVertexCount  float64        `db:"-" gorm:"-:all" json:"histAvgVertexCount,omitempty"`
	HistAvgTriangles    float64        `db:"-" gorm:"-:all" json:"histAvgTriangles,omitempty"`
	TestProtocolHistory []TestProtocol `db:"-" gorm:"-:all" json:"testProtocolHistory,omitempty"`
}

File: frontend/src/types/test.protocol.ts
import IProtocolEntryData from './protocol.entry';
import { TestResultState } from './test.result.state.enum';
import IDeviceData from './device';
import IProtocolPerformanceEntryData from './protocol.performance.entry';
import moment from 'moment';
import ITestRunData from "./test.run";

export default interface ITestProtocolData {
    id?: number | null,
    testRun: ITestRunData,
    testRunId: number,
    parentTestProtocolId: number | null,
    deviceId?: number | null,
    device?: IDeviceData | null,
    testName: string,
    startedAt: Date,
    endedAt?: Date,
    entries: IProtocolEntryData[]
    testResult: TestResultState
    performance: IProtocolPerformanceEntryData[]
    avgFps: number,
    avgMem: number,
    avgCpu: number,
    avgVertexCount: number,
    avgTriangles: number,
    histAvgFps: number,
    histAvgMem: number,
    histAvgCpu: number,
    histAvgVertexCount: number,
    histAvgTriangles: number,
    testProtocolHistory: ITestProtocolData[] | null,
    createdAt: Date,
    updatedAt: Date,
    deletedAt: Date,

    childProtocols: ITestProtocolData[],
}

export const duration = (startedAt: Date, endedAt: Date | undefined | null): string => {
    if (endedAt !== null && endedAt !== undefined) {
        const start = new Date(startedAt);
        const end = new Date(endedAt);
        const duration = end.valueOf() - start.valueOf();
        const m = moment.utc(duration);
        const secs = duration / 1000;
        if (secs > 60 * 60) {
            return m.format('h') + 'Std ' + m.format('m') + 'Min ' + m.format('s') + 'Sec';
        }
        if (secs > 60) {
            return m.format('m') + 'Min ' + m.format('s') + 'Sec';
        }
        return m.format('s') + 'Sec';
    }
    return 'running';
};
